library/modules/EventManager.cpp

```diff
@@ -597,57 +630,71 @@ static void manageJobCompletedEvent(color_ostream& out) {
+                JobCompleteData& prevJob = *prevIt;
+                if (prevJob.flags_bits_repeat && prevJob.completion_timer == 0
+                        && nowIt->completion_timer == -1) {
+                    df::job* seenJob = seenJobs[prevJob.id];
+                    // still false positive if cancelled at EXACTLY the right time, but experiments show this doesn't happen
+                    for (auto& [_, handle] : copy) {
+                        DEBUG(log, out).print("calling handler for repeated job completed event\n");
+                        run_handler(out, EventType::JOB_COMPLETED, handle, (void*)seenJob);
+                    }
+                }
+                ++prevIt;
+                ++nowIt;
+            } else { // prevJob.id > nowJob.id. Should never reach here.
+                ++nowIt;
             }
-            continue;
-        }
-
-        //recently finished or cancelled job
-        df::job& job0 = *prevJob.second;
-        if ( job0.flags.bits.repeat || job0.completion_timer != 0 )
-            continue;
-
-        for (auto &[_,handle] : copy) {
-            DEBUG(log,out).print("calling handler for job completed event\n");
-            run_handler(out, EventType::JOB_COMPLETED, handle, (void*) &job0);
         }
-    }
 
-    //erase old jobs, copy over possibly altered jobs
-    for (auto &[_,prev_job] : prevJobs) {
-        Job::deleteJobStruct(prev_job, true);
-    }
-    prevJobs.clear();
+        // Clean up garbage, if any.
+        if (seenJobs.size() > nowJobs.size()) {
+            struct CompareJobId {
+                bool operator()(const JobCompleteData& job, int32_t id) const {
+                    return job.id < id;
+                }
 
-    //create new jobs
-    for (auto &[_,now_job] : nowJobs) {
-        /*map<int32_t, df::job*>::iterator i = prevJobs.find((*j).first);
-        if ( i != prevJobs.end() ) {
-            continue;
-        }*/
+                bool operator()(int32_t id, const JobCompleteData& job) const {
+                    return id < job.id;
+                }
+            };
 
-        df::job* newJob = Job::cloneJobStruct(now_job, true);
-        prevJobs.emplace(newJob->id, newJob);
+            std::vector<int32_t> jobIdsGone;
+            for (auto& seenJob : seenJobs) {
+                if (!std::binary_search(nowJobs.begin(), nowJobs.end(), seenJob.first, CompareJobId{})) {
+                    jobIdsGone.push_back(seenJob.first);
+                }
+            }
+            for (auto jobId : jobIdsGone) {
+                //out.print("SEEN0: %d, %ld, %ld, %d, %ld, %d %s\n", jobId, seenJobs[jobId]->items.size(), seenJobs[jobId]->job_items.elements.size(), seenJobs[jobId]->flags.bits.working, seenJobs[jobId]->general_refs.size(), getWorkerID(seenJobs[jobId]), ENUM_ATTR(job_type, caption, seenJobs[jobId]->job_type));
+                Job::deleteJobStruct(seenJobs[jobId], true);
+                seenJobs.erase(jobId);
+            }
```

Comment:
I would be surprised if this is faster than iterating over `nowJobs` and using `seenJob.find` That's O(n) vs O(n log n). Maybe this is where your comparison time is being spent.

Reply:
My understanding is that we never call any handlers, unless tick1 > tick0. If I understand correctly, registering jobs that are seen while tick0=tick1, only adds detection of jobs that were queued (or modified) while the game was paused, having a `completion_timer` of 0, and that are already gone by the next job handler cycle. I'm having a hard time coming up with scenarios other than loading a save game, where this happens. Maybe initialization after loading can be handled differently. @myk002 What's your view on this?

Reply:
> My understanding is that we never call any handlers, unless tick1 > tick0

I think this is true, as is the rest of your analysis.

I believe the only other case of jobs appearing between ticks is when DFHack adds them, and we're not likely to delete those same jobs before the next tick.

IOW, I don't think we need sub-tick resolution for job detection.

Reply:
I think we need to distinguish between ticks and cycles of the event handler. That being said, if someone registers a handler for `jobCompleted` with a frequency of 150, they should already not rely on being notified for every job. 

As far as I understand, the difference between updating `seenJobs` or not when tick0=tick1is only relevant when there is a registered handler with frequency 0, otherwise the function is not going to be called anyway (except possibly during loading/initialization, not sure). This is ensured by the following guard in `manageEvents`:
```c++
 if ( tick >= eventLastTick[a] && tick - eventLastTick[a] < eventFrequency )
            continue;
```
Without changing documented behavior, I think we actually need option (2). I'm not sure whether there are any tools inside DFHack that actually register a job completed handle with frequency 0. 

I don't know what is calling `manageEvents`, so I don't know whether it is possible that it runs more than once per tick while the game is unpaused. If this can only happen while the game as paused, I suppose we can relax, because this will not impact (simulation) framerate.